core/src/main/scala/org/apache/spark/util/io/ChunkedByteBuffer.scala

```diff
@@ -84,6 +92,64 @@ private[spark] class ChunkedByteBuffer(var chunks: Array[ByteBuffer]) {
     }
   }
 
+  /**
+   * write to stream with zero copy if possible
+   */
+  def writeToStream(out: OutputStream): Unit = {
+    var buffer: Array[Byte] = null
+    val bufferLen = 1024 * 1024
+    writeBufferToDest(this, out.write)
+  }
+
+  /**
+   * write to ObjectOutput with zero copy if possible
+   */
+  override def writeExternal(out: ObjectOutput): Unit = {
+    // we want to keep the chunks layout
+    out.writeInt(chunks.length)
+    chunks.foreach(buffer => out.writeInt(buffer.limit()))
+    chunks.foreach(buffer => out.writeBoolean(buffer.isDirect))
+    writeBufferToDest(this, out.write)
+  }
+
+  override def readExternal(in: ObjectInput): Unit = {
+    val chunksNum = in.readInt()
+    val indices = 0 until chunksNum
+    val chunksSize = indices.map(_ => in.readInt())
+    val chunksDirect = indices.map(_ => in.readBoolean())
```

Comment:
Do we have cases where we want to preserve whether the buffer was direct or not across VM ? The current usecase does not require it ?
+CC @Ngone51 ?

If not, drop this and simplify the impl ?

Reply:
> Do we have cases where we want to preserve whether the buffer was direct or not across VM ?

I don't have such cases in my mind..

Reply:
Shall we default to on-heap buffer in deserialization regardless what it was before serialization? @mridulm @Ngone51

Reply:
Yes, let us do that - if some specific reason comes up in future, we can add support as required.
Will make the implementation much more simpler.

Reply:
Thanks for the suggestion, updated.

Reply:
Looks like we are still preservin `chunksDirect`, remove it ? And always deserialize to heap buffer via `ByteBuffer.wrap` ?